src/ripple/protocol/Feature.h

```diff
@@ -32,119 +32,63 @@
 
-namespace detail {
-
-// *NOTE*
-//
-// Features, or Amendments as they are called elsewhere, are enabled on the
-// network at some specific time based on Validator voting.  Features are
-// enabled using run-time conditionals based on the state of the amendment.
-// There is value in retaining that conditional code for some time after
-// the amendment is enabled to make it simple to replay old transactions.
-// However, once an Amendment has been enabled for, say, more than two years
-// then retaining that conditional code has less value since it is
-// uncommon to replay such old transactions.
-//
-// Starting in January of 2020 Amendment conditionals from before January
-// 2018 are being removed.  So replaying any ledger from before January
-// 2018 needs to happen on an older version of the server code.  There's
-// a log message in Application.cpp that warns about replaying old ledgers.
-//
-// At some point in the future someone may wish to remove Amendment
-// conditional code for Amendments that were enabled after January 2018.
-// When that happens then the log message in Application.cpp should be
-// updated.
-
-class FeatureCollections
-{
-    static constexpr char const* const featureNames[] = {
-        "MultiSign",      // Unconditionally supported.
-        "TrustSetAuth",   // Unconditionally supported.
-        "FeeEscalation",  // Unconditionally supported.
-        "OwnerPaysFee",
-        "PayChan",
-        "Flow",  // Unconditionally supported.
-        "CompareTakerFlowCross",
-        "FlowCross",
-        "CryptoConditions",
-        "TickSize",
-        "fix1368",
-        "Escrow",
-        "CryptoConditionsSuite",
-        "fix1373",
-        "EnforceInvariants",
-        "SortedDirectories",
-        "fix1201",
-        "fix1512",
-        "fix1513",
-        "fix1523",
-        "fix1528",
-        "DepositAuth",
-        "Checks",
-        "fix1571",
-        "fix1543",
-        "fix1623",
-        "DepositPreauth",
-        "fix1515",
-        "fix1578",
-        "MultiSignReserve",
-        "fixTakerDryOfferRemoval",
-        "fixMasterKeyAsRegularKey",
-        "fixCheckThreading",
-        "fixPayChanRecipientOwnerDir",
-        "DeletableAccounts",
-        // fixQualityUpperBound should be activated before FlowCross
-        "fixQualityUpperBound",
-        "RequireFullyCanonicalSig",
-        "fix1781",  // XRPEndpointSteps should be included in the circular
-                    // payment check
-        "HardenedValidations",
-        "fixAmendmentMajorityCalc",  // Fix Amendment majority calculation
-        "NegativeUNL",
-        "TicketBatch",
-        "FlowSortStrands",
-        "fixSTAmountCanonicalize",
-        "fixRmSmallIncreasedQOffers",
-    };
-
-    std::vector<uint256> features;
-    boost::container::flat_map<uint256, std::size_t> featureToIndex;
-    boost::container::flat_map<std::string, uint256> nameToFeature;
+enum class DefaultVote : bool { abstain = false, yes };
```

Comment:
My understanding is that it's the opposite - there is no such thing as a `no` vote. A validator either votes yes by sending the appropriate message, or it doesn't vote, which I think of as an abstention.

Reply:
This is a good discussion to have.

I see your point.  You're looking at the action of sending a message or the "lack of action" of not sending a message.  The action votes yes.  The lack of action, which you call abstaining, votes no.

I'm looking from a different angle, a perspective where we look at the consequences of the actions.  The current system of amendment voting mandates that all of the servers on the UNL take a position.  Today on the main net that's 41 servers.  Actively sending a vote says, yes, I support the amendment.  Not sending the vote has the effect of casting a no vote.  Any server on the UNL that does not send a vote has the _consequence_ of casting a no vote.

I see abstaining as a different animal.  If I abstain, then I'm saying that I am not participating in the vote.  I either don't care enough to have an opinion, or I trust those who are actively voting to be better informed than I am.  But in either case when I abstain I'm choosing to not participate in the vote.

If we were to implement something like abstaining it would have the effect of reducing the total number of votes on a particular amendment.  If one validator abstained from voting on an amendment then the number of required yes votes would change.

The currently implemented amendment voting system does not allow a validator to abstain.  Every operator of a validator on the UNL takes a position on every amendment, whether the operator wants to or not.  But I could imagine a future where something like abstaining might be useful.  I'd like to save the word "abstain" in case it becomes useful in the future.  Right now I think `yes` and `no` are good terms to reflect the consequences of the choice.

Does that makes sense?  I'm open to further discussion.

Reply:
I gotta admit, that really does make sense.